psi4/src/psi4/dlpno/mp2.cc

```diff
@@ -0,0 +1,1412 @@
+        // RMS of residual per LMO pair, for assessing convergence
+        std::vector<double> R_iajb_rms(n_lmo_pairs, 0.0);
+
+// Calculate residuals from current amplitudes
+#pragma omp parallel for schedule(static, 1)
+        for (int ij = 0; ij < n_lmo_pairs; ++ij) {
+            int i, j;
+            std::tie(i, j) = ij_to_i_j[ij];
+
+            R_iajb[ij] = std::make_shared<Matrix>("Residual", n_pno[ij], n_pno[ij]);
+
+            if (n_pno[ij] == 0) continue;
+
+            for (int a = 0; a < n_pno[ij]; ++a) {
+                for (int b = 0; b < n_pno[ij]; ++b) {
+                    R_iajb[ij]->set(a, b,
+                                    K_iajb[ij]->get(a, b) +
+                                        (e_pno[ij]->get(a) + e_pno[ij]->get(b) - F_lmo->get(i, i) - F_lmo->get(j, j)) *
+                                            T_iajb[ij]->get(a, b));
+                }
+            }
+
+            for (int k = 0; k < naocc; ++k) {
+                int kj = i_j_to_ij[k][j];
+                int ik = i_j_to_ij[i][k];
+
+                if (kj != -1 && i != k && fabs(F_lmo->get(i, k)) > options_.get_double("F_CUT") && n_pno[kj] > 0) {
+                    SharedMatrix temp =
+                        linalg::triplet(S_pno_ij_kj[ij][k], T_iajb[kj], S_pno_ij_kj[ij][k], false, false, true);
+                    temp->scale(-1.0 * F_lmo->get(i, k));
+                    R_iajb[ij]->add(temp);
+                }
+                if (ik != -1 && j != k && fabs(F_lmo->get(k, j)) > options_.get_double("F_CUT") && n_pno[ik] > 0) {
+                    SharedMatrix temp =
+                        linalg::triplet(S_pno_ij_ik[ij][k], T_iajb[ik], S_pno_ij_ik[ij][k], false, false, true);
+                    temp->scale(-1.0 * F_lmo->get(k, j));
+                    R_iajb[ij]->add(temp);
+                }
+            }
+
+            R_iajb_rms[ij] = R_iajb[ij]->rms();
+        }
+
+        // evaluate convergence using current amplitudes and residuals
+        e_prev = e_curr;
+        e_curr = eval_amplitudes();
+        r_curr = *max_element(R_iajb_rms.begin(), R_iajb_rms.end());
+
+        r_converged = (fabs(r_curr) < options_.get_double("R_CONVERGENCE"));
+        e_converged = (fabs(e_curr - e_prev) < options_.get_double("E_CONVERGENCE"));
+
+// use residuals to get next amplitudes
+#pragma omp parallel for schedule(static, 1)
+        for (int ij = 0; ij < n_lmo_pairs; ++ij) {
+            int i, j;
+            std::tie(i, j) = ij_to_i_j[ij];
+            for (int a = 0; a < n_pno[ij]; ++a) {
+                for (int b = 0; b < n_pno[ij]; ++b) {
+                    T_iajb[ij]->add(a, b, -R_iajb[ij]->get(a, b) / ((e_pno[ij]->get(a) + e_pno[ij]->get(b)) -
+                                                                    (F_lmo->get(i, i) + F_lmo->get(j, j))));
+                }
+            }
+        }
+
+        // DIIS extrapolation
+        SharedMatrix T_iajb_flat = flatten_mats(T_iajb);
```

Comment:
I know this is picky, but I'd prefer you flatten to a Vector rather than a Matrix. This object is inherently 1D.

Reply:
I would also have liked to use a `Vector` here, unfortunately the DIIS machinery expects a `Matrix`: https://github.com/psi4/psi4/blob/master/psi4/src/psi4/libdiis/diismanager.h

Reply:
Are you sure? If you use [the variadic functions](https://github.com/psi4/psi4/blob/master/psi4/src/psi4/libdiis/diismanager.h#L70-L73), you can accept an arbitrary number of arguments and even different DIISEntry type. Example usage [here](https://github.com/psi4/psi4/blob/master/psi4/src/psi4/dct/dct_oo_UHF.cc#L69-L71]). The [DIISEntry::Vector exists](https://github.com/psi4/psi4/blob/master/psi4/src/psi4/libdiis/diisentry.h#L55), so vectors should work.

Reply:
Good to know, those variadic functions are super neat. I changed the code to use `DIISEntry::Vector`